### PR TITLE
Design doc for Ubernetes Lite

### PR DESCRIPTION
**Pull request by justinsb**
Documentation of the Ubernetes-Lite idea and the main design & implementation
points.

Not sure if we should include the GCE vs AWS complexities in this doc - what do you think @quinton-hoole ?  (i.e. how AWS has a single API for multiple zones, but GCE has one per-zone, and then the naming scope issues)

### CODE REVIEW

**Comment by quinton-hoole** [comment]
Thanks Justin.  Yes, I'd suggest outlining cloud-specific implementation details, and adding links to relevant PR's.  I'll review more fully shortly.

File: docs/proposals/federation-lite.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+## Design
+
+These are the main requirements:
+
+1. kube-up must allow bringing up a cluster that spans multiple zones.
+1. pods in a replication controller should attempt to spread across zones.
+1.  kubernetes should not schedule pods which require volumes to hosts in a
+different zone.
+
```
Comment (quinton-hoole) [inline docs/proposals/federation-lite.md]: I think it's worth discussing briefly how cross-zone load balancing will work.  I believe that it "just works" due to both GCE and AWS regional load balancing, but that fact is conspicuously absent from this design, and should be added.

**Comment by quinton-hoole** [comment]
Other than my comments, LGTM.

File: docs/proposals/federation-lite.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+volumes.  We will create an admission controller `PersistentVolumeLabel`.
+`PersistentVolumeLabel` will intercept requests to create PersistentVolumes,
+and will label them appropriately by calling in to the cloud provider.
+
+## AWS Specific Considerations
+
+The AWS implementation here is fairly straightforward.  The AWS API is
+region-wide, meaning that a single call will find instances and volumes in all
+zones.  In addition, instance ids and volume ids are unique per-region (not
+per-zone).  I believe they are actually globally unique, but I do not know if
```
Comment (quinton-hoole) [inline docs/proposals/federation-lite.md]: The wording here is slightly ambiguous (uninformed readers might assume that id's are not unique per zone).  Also, it's names that are not necessarily unique, right.  ID's are GUIDS, so always globally unique.  Suggest rewording

**Comment by quinton-hoole** [comment]
Thanks for the updates @justinsb.  LGTM!

**Comment by quinton-hoole** [comment]
Oh, just the one minor rewording suggestion @justinsb.  I'll remove the LGTM label so that merge-bot doesn't merge this until that small change is in.

File: docs/proposals/federation-lite.md (hunk 2 above)
Comment (justinsb) [inline docs/proposals/federation-lite.md]: Good call.  I reworded to:

`In addition, instance ids and volume ids are unique per-region (and
hence also per-zone).  (They may be globally unique also - but we do not need
a global-uniqueness guarantee because Ubernetes-Lite will not span regions.)`

AWS instance IDs look like `i-456789ab` or similar - they're not GUIDs.  Best documentation I've ever seen on their structure is http://www.jackofallclouds.com/2009/09/anatomy-of-an-amazon-ec2-resource-id/, although I think AWS treats them as opaque tokens.

**Comment by justinsb** [comment]
Repushed with that wording changed to be less misleading!

**Comment by quinton-hoole** [comment]
@justinsb Cool!  LGTM.

**Comment by quinton-hoole** [comment]
flakey e2e test tracked in #19304.

@k8s-bot unit test this

**Comment by quinton-hoole** [comment]
A known flaky test tracked elsewhere.  Try again.

@k8s-bot test this

**Comment by quinton-hoole** [comment]
It's not at all clear what went wrong with the unit and integration tests.  Trying again while I dig further.

@k8s-bot unit test this
